src/lowlevel/writer_buffered.rs

```diff
@@ -29,20 +33,26 @@ impl<W: AsyncWrite + Unpin> WriterBuffered<W> {
     /// # Cancel Safety
     ///
     /// This function is cancel safe, but it might cause the data to be partially written.
-    pub(crate) async fn write_all(&mut self, buf: &[u8]) -> Result<(), io::Error> {
-        self.0.get_mut().write_all(buf).await
+    pub(crate) async fn write_all(self: Pin<&Self>, buf: &[u8]) -> Result<(), io::Error> {
```

Comment:
This whole setup still feels wrong to me. Shouldn't all these methods take `Pin<&mut Self>` instead, and then there would be no need for a lock in the first place?

Reply:
Since we cannot use`get_mut`, we would have to take a `Pin<&Self>` as it is stored in an `Arc`.

Reply:
Since we cannot useget_mut, we would have to take a Pin<&Self> as it is stored in an Arc.

Reply:
> Shouldn't all these methods take Pin<&mut Self> instead, and then there would be no need for a lock in the first place?

About the locking part, that is because we can have multiple `WriteEnd`s, so having mutex here is necessary but since we have a buffer sitting before it, the mutex probably won't be contented.

Reply:
Why is it possible to have multiple `WriteEnd`s? Is that just for convenience, or is it a feature of the underlying protocol? If it's for multiplexing, I'm guessing you didn't want to spawn a task that is responsible for doing the actual writes and have each `WriteEnd` just hold `mpsc::Sender`s instead because it requires allocation + copying bytes? Another alternative would be to make writing through a `WriteEnd` require a `&mut Connection` so that the onus is on the caller to add a lock if they need, rather than making it internal. So the API would be something like:

```rust
impl WriteEnd {
    async fn write(&mut self, bytes: &[u8], connection: &mut Connection) -> io::Result<usize> { ... }
}
```

Reply:
> Why is it possible to have multiple WriteEnds? Is that just for convenience, or is it a feature of the underlying protocol?

The underlying sftp protocol supports sending multiple requests at the same time, provided you give them different request ids.

> If it's for multiplexing, I'm guessing you didn't want to spawn a task that is responsible for doing the actual writes and have each WriteEnd just hold mpsc::Senders instead because it requires allocation + copying bytes? 

Yeah, pretty much sum it up.

I wanted to use a `Bytes` buffer so that I can efficiently store data from the individual buffer in `WriteEnd` and reuse them and arbitary `Bytes` created by the user.

> Another alternative would be to make writing through a WriteEnd require a &mut Connection so that the onus is on the caller to add a lock if they need, rather than making it internal.

That makes sense!
But that would definitely requires another PR.

Though I am also thinking about extracting the `Bytes` buffer out and refactor it to looks like a `mpsc`, that would make the internals

Reply:
Opened #9 to track this.